internal/api/ring.go

```diff
@@ -541,6 +558,76 @@ func handleRetryReleaseRing(c *Context, w http.ResponseWriter, r *http.Request)
 	outputJSON(c, w, ring)
 }
 
+// handlePauseReleaseRing responds to POST /api/rings/release/pause, pausing all pending releases
+func handlePauseReleaseRing(c *Context, w http.ResponseWriter, r *http.Request) {
+	ringsPending, err := c.Store.GetRingsInPendingState()
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to get all rings pending work")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	c.Logger.Info("pausing all releases in pending state. Releases already in progress cannot be paused")
+
+	for _, ring := range ringsPending {
+		ring.State = model.RingStateReleasePaused
+	}
+
+	c.Logger.Debug("Updating all rings in a single transaction")
+	if err = c.Store.UpdateRings(ringsPending); err != nil {
+		c.Logger.WithError(err).Error("failed to update rings status to paused in a single transaction")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+}
+
+// handleResumeReleaseRing responds to POST /api/rings/release/pause, pausing all pending releases
+func handleResumeReleaseRing(c *Context, w http.ResponseWriter, r *http.Request) {
+	ringsPaused, err := c.Store.GetRingsInPendingState()
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to get all rings in paused state")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	c.Logger.Info("resuming all releases in paused state")
+
+	for _, ring := range ringsPaused {
+		ring.State = model.RingStateReleasePending
+	}
+
+	c.Logger.Debug("Updating all rings in a single transaction")
+	if err = c.Store.UpdateRings(ringsPaused); err != nil {
+		c.Logger.WithError(err).Error("failed to update rings status to pending in a single transaction")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+}
+
+// handleCancelReleaseRing responds to POST /api/rings/release/pause, pausing all pending releases
+func handleCancelReleaseRing(c *Context, w http.ResponseWriter, r *http.Request) {
+	ringsPending, err := c.Store.GetRingsInPendingState()
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to get all rings in pending state")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	c.Logger.Info("canceling all releases in pending state. Setting desired release same as active. Releases already in progress cannot be cancelled")
+
+	for _, ring := range ringsPending {
+		ring.State = model.RingStateStable
+		ring.DesiredReleaseID = ring.ActiveReleaseID
```

Comment:
Would it be a good idea to have some state that shows the last release was cancelled in case we run in to issues in future releases due to undone work?

Reply:
We do have active release and desired release. So I think it clear on what is running where when we request a new ring release. Does this cover you question?